java/src/main/java/com/google/appengine/tools/pipeline/impl/model/Slot.java

```diff
@@ -40,68 +43,74 @@
 
   // persistent
   private boolean filled;
-  private Date fillTime;
+  private Instant fillTime;
   private Object value;
   private Key sourceJobKey;
   private final List<Key> waitingOnMeKeys;
 
   // transient
   private List<Barrier> waitingOnMeInflated;
-  private Object serializedVersion;
+  //either a Blob or List<Key> ... given that we know implementation details of serialization
+  private Object serializedVersion; //TODO: 'guaranteed' to be type serializable by datastore; what does this mean??
+  private SerializationStrategy serializationStrategy;
 
-
-  public Slot(Key rootJobKey, Key generatorJobKey, String graphGUID) {
+  public Slot(Key rootJobKey, Key generatorJobKey, String graphGUID, SerializationStrategy serializationStrategy) {
     super(rootJobKey, generatorJobKey, graphGUID);
-    waitingOnMeKeys = new LinkedList<>();
+    this.waitingOnMeKeys = new LinkedList<>();
+    this.serializationStrategy = serializationStrategy;
   }
 
-  public Slot(Entity entity) {
-    this(entity, false);
+  public Slot(Entity entity, SerializationStrategy serializationStrategy) {
+    this(entity, serializationStrategy, false);
+    this.serializationStrategy = serializationStrategy;
   }
 
-  public Slot(Entity entity, boolean lazy) {
+  public Slot(Entity entity, SerializationStrategy serializationStrategy, boolean lazy) {
     super(entity);
-    filled = (Boolean) entity.getProperty(FILLED_PROPERTY);
-    fillTime = (Date) entity.getProperty(FILL_TIME_PROPERTY);
-    sourceJobKey = (Key) entity.getProperty(SOURCE_JOB_KEY_PROPERTY);
+    filled = entity.getBoolean(FILLED_PROPERTY);
+    fillTime = EntityUtils.getInstant(entity, FILL_TIME_PROPERTY);
+    sourceJobKey = entity.getKey(SOURCE_JOB_KEY_PROPERTY);
     waitingOnMeKeys = getListProperty(WAITING_ON_ME_PROPERTY, entity);
+    serializationStrategy = serializationStrategy;
     if (lazy) {
-      serializedVersion = entity.getProperty(VALUE_PROPERTY);
+      serializedVersion = entity.getBlob(VALUE_PROPERTY);
     } else {
-      value = deserializeValue(entity.getProperty(VALUE_PROPERTY));
+      value = deserializeValue(entity.getBlob(VALUE_PROPERTY));
     }
```

Comment:
all of the above subject to bugs, if there are cases where property may be missing, but we don't expect it